### DISCUSSION TITLE
Use versioned migrations (i.e. use atlas)

### DISCUSSION THREAD

**Discussion by nicolasfranck** [ugent-library/old-people-service#56] [Ideas]
The database schema is now automatically created by entgo, which uses the golang code
from atlas to inspect the current table schema, compare it to the desired schema, and then
apply the necessary changes to the database.

This happens every time the application starts up, and the postgres is made:

https://github.com/ugent-library/people-service/blob/main/repository/common.go#L89
```
txClient.Client().Schema.Create(context.Background())
```

But with atlas, one can also use versioned migrations. This means that there is a directory,
here called `migrations/` filled with sql files (prefixed with a timestamp for sorting). These
sql are created with the help of atlas, but can be manually edited.

Pending migrations are executed by atlas as such:

```
atlas migrate apply --url "$PEOPLE_DB_URL" --dir "file://migrations"
```

Atlas uses its own table `atlas_schema_revisions` stored in schema `atlas_schema_revisions`
to track which migrations are pending, and which have been executed already.

One "can" also apply these migrations using the atlas golang api:

```
db, err := sql.Open("pgx", config.Db.Url)
if err != nil {
	return err
}

dir, err := migrate.NewLocalDir("migrations")
if err != nil {
	return err
}

driver, err := postgres.Open(db)
if err != nil {
	return err
}

executor, err := migrate.NewExecutor(
	driver,
	dir,
	migrate.NopRevisionReadWriter{},
	// why is this needed? "atlas migrate" runs fine without this!
	migrate.WithBaselineVersion("20230919120247"),
)
```

See https://github.com/ugent-library/people-service/blob/atlas/cli/migrate.go for full code example

But as you can see, it uses the no-op struct `migrate.NopRevisionReadWriter` which does
basically nothing, and therefore does not inspect that revision table. So if you execute it
without providing a baseline version, it will return an error stating that the database is not clean.

With that baseline version (the first file here) it works for now, but because
the revision table is not read, it will in the future always execute all sql files AFTER
that base revision file.

According [this comment](https://github.com/ariga/atlas/issues/1527#issuecomment-1613835851) the
implementing RevisionReadWriter for this is located in an internal package (and used by the command `atlas migrate`).
See also https://github.com/ariga/atlas/blob/160fc1d9981e31be0f1a34b8ff90a6e95d0a990c/cmd/atlas/internal/migrate/migrate.go#L54

With this, we have to bring the full atlas command (70MB) when deploying applications with atlas.
Without this full command, the binary "people-service" is now 30MB, with some atlas functionality.
That would make 100MB

**Comment by netsensei** [comment]
Two thoughts on this:

First, we should re-think if we even want to actively generate the database schema on first boot of the application. Ideally, we separate that as a concern from the application proper. The latter should only check whether or not an initialized database is there (e.g. a simple check for presence of tables, etc.) but not take any further action.

Second, in the same vain: do we want to incorporate atlas as part of the application proper? I don't think so, as automatically running migrations on application boot does come with a few issues e.g. what if a migration contains a mistake? what if we want to do a dry-run first? how do we deal with data migrations next to schema migrations? etc.

If anything, I would argue to keep database management entirely separate from the application if we can.

**Reply by nicolasfranck** [reply]
@netsensei In an ideal situation, yes. But in theory a migration might introduce a breaking change that the older code cannot work with, while the newer code cannot work with the old database schema. For this reason applications like rails requires the application to ship the migration along with the application and upgrade it before restarting. If we separate it, all table changes require a manual downtime, and we have to remember it, as the code will not check for compatibility until it fails.

**Reply by netsensei** [reply]
Doesn't Rails have logic to manage migrations separate from the application proper through the `rake db:migrate` command?

**Reply by nics** [reply]
yes, the migration code lives with the application and the migration is called through a task manager

**Reply by nicolasfranck** [reply]
@netsensei yes, but it is still shipped with the application. 

Anyway, even if we put the migration out of the application itself, the application itself needs to aware that is it working on an incompatible version. Rails solves this by throwing an error "Pending migrations". That is a must-have if we're planning on splitting this functionality.

**Reply by netsensei** [reply]
To my mind, the issue here is shipping `atlas` as an integrated part of the application binary in order to be able to execute migrations from within the docker container on the cluster. I have strong doubts whether that's the way forward: we ought to research what best practices are regarding managing database & data migrations within a cloud infra context.

I do think it makes sense to keep the migration files themselves as part of the codebase as it lives on github for obvious reasons.

**Reply by nicolasfranck** [reply]
@netsensei agreed. Embedding atlas within the application may be problematic (well through entgo some part still is there). The only examples I can find is where they actually use the atlas binary to run the migrations before restarting the application. I guess rolling updates within nomad are problematic too: one container spins up, is not yet available, but it runs a database migration at first, which directly influences other instances that are about to be replaced. But that goes for the current table upgrades too.

**Reply by netsensei** [reply]
@nicolasfranck some documentation:

* https://itnext.io/database-migrations-on-kubernetes-using-helm-hooks-fb80c0d97805
* https://codefresh.io/blog/database-migrations-in-the-era-of-kubernetes-microservices/
* https://danielabaron.me/blog/nomad-tips-and-tricks/

The last blogpost shows how you can use Nomad's `lifecycle` stanza to run a task during `prestart` before the main tasks is started.
